src/library/tactic/simp_util.h

```diff
@@ -40,59 +40,72 @@ class instantiate_emetas_fn {
         m_prover(prover) {}
 
     bool operator()(tmp_type_context & tmp_ctx, list<expr> const & emetas, list<bool> const & instances) {
-        bool failed = false;
-        for_each2(emetas, instances, [&](expr const & mvar, bool const & is_instance) {
-                unsigned mvar_idx = to_meta_idx(mvar);
-                if (failed) return;
-                expr mvar_type = tmp_ctx.instantiate_mvars(tmp_ctx.infer(mvar));
-                if (has_idx_metavar(mvar_type)) {
-                    failed = true;
-                    return;
-                }
+        // Repeat until we stop making progress or all variables are instantiated.
+        bool progress = false;
+        bool done = false;
+        do {
+            progress = false; // Set to true if we instantiate a variable.
+            done = true; // Set to false if we skip a variable.
+            for_each2(emetas, instances, [&](expr const & mvar, bool const & is_instance) {
+                    unsigned mvar_idx = to_meta_idx(mvar);
+                    expr mvar_type = tmp_ctx.instantiate_mvars(tmp_ctx.infer(mvar));
 
-                if (tmp_ctx.is_eassigned(mvar_idx)) return;
+                    if (tmp_ctx.is_eassigned(mvar_idx)) return;
 
-                if (is_instance) {
-                    if (auto v = tmp_ctx.ctx().mk_class_instance(mvar_type)) {
-                        if (!tmp_ctx.is_def_eq(mvar, *v)) {
+                    if (is_instance && tmp_ctx.ctx().ready_to_synthesize(mvar_type)) {
+                        // Use the *temporary* context for the instance search
+                        // in order to also fill in any previous temporary metas created by matching.
+                        if (auto v = tmp_ctx.mk_class_instance(mvar_type)) {
+                            if (!tmp_ctx.is_def_eq(mvar, *v)) {
+                                lean_simp_trace(tmp_ctx, name({"simplify", "failure"}),
+                                                tout() << "unable to assign instance for: " << mvar_type << "\n";);
+                                done = false;
+                                return;
+                            } else {
+                                progress = true;
+                            }
+                        } else {
                             lean_simp_trace(tmp_ctx, name({"simplify", "failure"}),
-                                            tout() << "unable to assign instance for: " << mvar_type << "\n";);
-                            failed = true;
+                                            tout() << "unable to synthesize instance for: " << mvar_type << "\n";);
+                            done = false;
```

Comment:
I think we can do an early return here now.  If it fails once, then it's not going to succeed in a future iteration (because of the ready-to-synthesize check).  The same goes for the auto-param/discharging code below, if it fails we shouldn't try to fill in the other parameters.  (I'm a bit concerned about performance here. Ideally we'd only try to synthesize each argument once, and fail immediately without futilely solving the rest if it doesn't succeed.)

Reply:
I wanted to retry the instances to avoid having cases like `{α : Type*} [some_class_with_out_param α] [some_class_with_in_param α]` work and `{α : Type*} [some_class_with_in_param α] [some_class_with_out_param α]` fail. As long as it's guarded by the metavariable check, I agree that this should work.